### PR TITLE
Add Getter.Keys() with Jaeger Baggage support.

### LINKED ISSUE #967: Add Jaeger propagator
Similar to https://github.com/open-telemetry/opentelemetry-java/issues/853 but for Jaeger clients.

* https://www.jaegertracing.io/docs/1.17/client-libraries/#propagation-format
* https://github.com/jaegertracing/jaeger-client-java/blob/master/jaeger-core/src/main/java/io/jaegertracing/internal/propagation/TextMapCodec.java

Does OTEL client support exporting/injecting of TraceContext/SpanContext into two formats? At the same time use W3C TraceContext and Jaeger? The use case is to allow gradual migration of existing deployments using Jaeger prop. format to W3C.

### PR DESCRIPTION
**Pull request by carlosalberto**
Fixes #967 

Implements the https://github.com/open-telemetry/opentelemetry-specification/pull/825 update, and implements Jaeger Baggage support to verify that the API is correct/useful (based on [their own propagator](https://github.com/jaegertracing/jaeger-client-java/blob/master/jaeger-core/src/main/java/io/jaegertracing/internal/propagation/TextMapCodec.java))

- Used `Iterable` for `keys()`.
- Had to remove the usage of anonymous classes in the integration test's `trace-context` section, as it broke some versions of Java 11 (see [this](https://stackoverflow.com/questions/54775253/jdk-11-0-2-compilation-fails-with-javac-npe-on-anonymous-parameterized-class-typ)).
- We can play with removing the (most likely) slow regex in a follow-up.
- Don't put an empty `Baggage` in `Context` in case there was not Baggage to be extracted at all.

For Jaeger reviewers:

- Didn't encode the Baggage values. In Jaeger there's two `TextMap` propagators registered by default (encoded and non-encoded propagation). We need to define the encoding case in a follow-up (somehow).

@open-telemetry/java-approvers 
cc @pavolloffay @jpkrohling

### CODE REVIEW

**Comment by richardstartin** [comment]
The problem with making this a collection is it will often need to be materialised, even if you can get away with calling `Map.keySet` in a lot of cases. Making it an `Iterable` would be lighter-weight in most cases, e.g. Kafka client, old enumeration based context carriers (e.g. JMS, `HttpServletRequest`)

**Comment by carlosalberto** [comment]
I'd be fine to have `keys()` return an `Iterable` or `Iterator`.

**Comment by jkwatson** [comment]
Can you add the concrete example of how the Jaeger propagator will use this new feature to this PR? I think that will be very useful for the spec PR, as well, to demonstrate how you're doing to use it.

File: context_prop/src/main/java/io/opentelemetry/context/propagation/TextMapPropagator.java [hunk 1]
```diff
@@ -115,6 +116,14 @@
    */
   interface Getter<C> {
 
+    /**
+     * Returns all the keys in the given carrier.
+     *
+     * @param carrier carrier of propagation fields, such as an http request.
```
Comment (jkwatson) [inline context_prop/src/main/java/io/opentelemetry/context/propagation/TextMapPropagator.java:122]:
```suggestion
     * @param carrier The carrier of propagation fields, such as an http request.
```

File: context_prop/src/main/java/io/opentelemetry/context/propagation/TextMapPropagator.java [hunk 2]
```diff
@@ -115,6 +116,14 @@
...
+     * @since 0.10.0
```
Comment (jkwatson) [inline context_prop/src/main/java/io/opentelemetry/context/propagation/TextMapPropagator.java:123]: I think we've gotten rid of all the `@since` tags for now; we'll set them all to 1.0 when we're ready to release.

File: extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java [hunk 3]
```diff
@@ -187,6 +208,42 @@ public static JaegerPropagator getInstance() {
...
+        }
+        builder.put(key.substring(BAGGAGE_PREFIX.length()), getter.get(carrier, key));
+      } else if (key.equals(BAGGAGE_HEADER)) {
+        builder = parseBaggageHeader(getter.get(carrier, key), builder);
+      }
+    }
+    return builder == null ? null : builder.build();
+  }
+
+  @SuppressWarnings("StringSplitter")
```
Comment (jkwatson) [inline extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java:231]: not relevant for this PR in particular, but I think we should turn off this warning globally, since it's telling us to use guava, which we don't to do for the API, and won't, ever.

File: extensions/trace_propagators/src/test/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagatorTest.java [hunk 4]
```diff
@@ -133,12 +145,49 @@ void inject_NotSampledContext() {
...
+                BAGGAGE_PREFIX + "meta", "meta-value"));
+  }
+
   @Test
   void extract_Nothing() {
     // Context remains untouched.
     assertThat(
             jaegerPropagator.extract(
-                Context.current(), Collections.<String, String>emptyMap(), Map::get))
+                Context.current(), Collections.<String, String>emptyMap(), getter))
```
Comment (jkwatson) [inline extensions/trace_propagators/src/test/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagatorTest.java:190]: might be a little safer/cleaner to just use `Context.root()` here, so we know that we don't have a pre-polluted context.

**Comment by jkwatson** [comment]
This looks ok (although it makes me sad that our Getter is no longer a SMI). Based on the discussion above, I had thought you would go with an Iterable/Iterator, rather Collection as the return type for `keys()`. Did something change your mind?

File: extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java (hunk 3 above)
Comment (carlosalberto) [reply extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java:231]: +1

File: context_prop/src/main/java/io/opentelemetry/context/propagation/TextMapPropagator.java (hunk 2 above)
Comment (carlosalberto) [reply context_prop/src/main/java/io/opentelemetry/context/propagation/TextMapPropagator.java:123]: Oh, I totally forgot about it. Will do.

File: extensions/trace_propagators/src/test/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagatorTest.java (hunk 4 above)
Comment (carlosalberto) [reply extensions/trace_propagators/src/test/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagatorTest.java:190]: Fair enough, will update.

**Comment by carlosalberto** [comment]
> I had thought you would go with an Iterable/Iterator, rather Collection as the return type for keys(). Did something change your mind?

Wanted to give a last try at `Collection` and to verify it's still fine ;) But if you think we should try enumerable instead (as mentioned in the previous related conversation), then I will go with enumerable for sure.

**Review by jpkrohling — COMMENTED** [review]
The Jaeger specific parts are looking good to me, but I'm not an expert in the propagation format. Perhaps @objectiser or @yurishkuro would want to double-check as well?

File: extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java [hunk 5]
```diff
@@ -108,16 +118,27 @@ public static JaegerPropagator getInstance() {
...
 
     SpanContext spanContext = getSpanContextFromHeader(carrier, getter);
-    if (!spanContext.isValid()) {
-      return context;
+    if (spanContext.isValid()) {
+      context = context.with(Span.wrap(spanContext));
+    }
+
+    Baggage baggage = getBaggageFromHeader(carrier, getter);
+    if (baggage != null) {
```
Comment (anuraaga) [inline extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java:137]: Does it make sense to add empty baggage similar to how we add invalid span?

File: extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java [hunk 6]
```diff
@@ -187,6 +208,42 @@ public static JaegerPropagator getInstance() {
     return buildSpanContext(traceId, spanId, flags);
   }
 
+  private static <C> Baggage getBaggageFromHeader(C carrier, Getter<C> getter) {
+    Baggage.Builder builder = null;
+
+    for (String key : getter.keys(carrier)) {
+      if (key.startsWith(BAGGAGE_PREFIX)) {
+        if (key.length() <= BAGGAGE_PREFIX.length()) {
```
Comment (anuraaga) [inline extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java:216]: Since we already checked startsWith I think this can be equals check

File: extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java [hunk 7]
```diff
@@ -187,6 +208,42 @@ public static JaegerPropagator getInstance() {
...
+      } else if (key.equals(BAGGAGE_HEADER)) {
+        builder = parseBaggageHeader(getter.get(carrier, key), builder);
+      }
+    }
+    return builder == null ? null : builder.build();
+  }
+
+  @SuppressWarnings("StringSplitter")
+  private static Baggage.Builder parseBaggageHeader(String header, Baggage.Builder builder) {
+    for (String part : header.split("\\s*,\\s*")) {
```
Comment (anuraaga) [inline extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java:233]: Regex split looks like it could be very slow. Best could be to use indexOf to avoid split array allocation but if too tedious can we at least just split on the single chars?

File: extensions/trace_propagators/src/test/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagatorTest.java [hunk 8]
```diff
@@ -313,6 +362,78 @@ void extract_UrlEncodedContext() throws UnsupportedEncodingException {
...
+    carrier.put(BAGGAGE_HEADER, "nometa+novalue");
+
+    assertThat(getBaggage(jaegerPropagator.extract(Context.current(), carrier, getter)))
+        .isEqualTo(Baggage.empty());
+  }
+
+  @Test
+  void extract_baggageOnly_withHeader_andPrefix() {
+    Map<String, String> carrier = new LinkedHashMap<>();
+    carrier.put(BAGGAGE_HEADER, "nometa=nometa-value,meta=meta-value");
```
Comment (anuraaga) [inline extensions/trace_propagators/src/test/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagatorTest.java:425]: Can you add some test cases with invalid, and with whitespace?

**Review by anuraaga — COMMENTED** [review]
Mostly looks nice thanks

**Comment by yurishkuro** [comment]
> In Jaeger there's two TextMap propagators registered by default (encoded and non-encoded propagation). Which one is the default? 

Neither is default. In OT, HTTP and TextMap are different, so Jaeger url-encodes baggage for HTTP, but not for TextMap (eg in Kafka headers there's no need to url-encode).

> The jaeger-baggage header support only happens at extraction, not injection, which makes me believe it's deprecated? Should we mention this in the docs?

That is correct. The official baggage format (in and out) is uberctx-{key}. jaeger-baggage header was invented later for manual injection via curl.

File: extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java (hunk 5 above)
Comment (carlosalberto) [reply extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java:137]: That's a tricky one (I will let @pavolloffay @jpkrohling comment), but I'd imagine we wanting to keep whatever `Baggage` we had before, in case there was no baggage *at all*.

File: extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java (hunk 7 above)
Comment (carlosalberto) [reply extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java:233]: Let's do that in a follow-up. This code is ported directly from the Jaeger Propagator, but yes, we could try to optimize things in follow ups (we did the same for OC's TraceContext propagator, for example).

File: extensions/trace_propagators/src/test/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagatorTest.java (hunk 8 above)
Comment (carlosalberto) [reply extensions/trace_propagators/src/test/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagatorTest.java:425]: Oh, yes, totally missed that.

File: context_prop/src/main/java/io/opentelemetry/context/propagation/TextMapPropagator.java [hunk 9]
```diff
@@ -115,6 +116,14 @@
    */
   interface Getter<C> {
 
+    /**
+     * Returns all the keys in the given carrier.
+     *
+     * @param carrier The carrier of propagation fields, such as an http request.
+     * @since 0.10.0
+     */
+    Collection<String> keys(C carrier);
```
[the remark this answers is not in the document]
Comment (carlosalberto) [reply context_prop/src/main/java/io/opentelemetry/context/propagation/TextMapPropagator.java:125]: Fair enough. Will update to go the iterable way.

**Comment by carlosalberto** [comment]
> Neither is default. In OT, HTTP and TextMap are different, so Jaeger url-encodes baggage for HTTP, but not for TextMap (eg in Kafka headers there's no need to url-encode).

Oh, interesting. @yurishkuro , what do you think we should do here? Maybe have two versions, one that does encoding, one that does not? Or maybe something else? e.g.

```
  private JaegerPropagator(bolean doBaggageEncoding) {
     this.doBaggageEncoding = doBaggageEncoding; // Depending on this, baggage will be encoded or not.
  }
```

File: extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java [hunk 10]
```diff
@@ -187,6 +208,42 @@ public static JaegerPropagator getInstance() {
...
+        if (key.length() <= BAGGAGE_PREFIX.length()) {
+          continue;
+        }
+
+        if (builder == null) {
+          builder = Baggage.builder();
+        }
+        builder.put(key.substring(BAGGAGE_PREFIX.length()), getter.get(carrier, key));
+      } else if (key.equals(BAGGAGE_HEADER)) {
+        builder = parseBaggageHeader(getter.get(carrier, key), builder);
```
Comment (yurishkuro) [inline extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java:225]: Jaeger SDKs would combine baggage from `uberctx-{key}` and `jaeger-baggage` headers, you're overriding it instead.

Comment (carlosalberto) [reply extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java:225]: Actually they are getting merged, i.e. see [this test](https://github.com/open-telemetry/opentelemetry-java/pull/1549/files#diff-4e39f9e60626e3859be5734ea004b64bbbc1530e9a3d8e685122164aebfe27cdR423) - let me know otherwise :)

File: extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java (hunk 6 above)
Comment (carlosalberto) [reply extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java:216]: I'd rather stick to check the length, as it only depends on a numerical comparison.

File: extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java (hunk 10 above)
Comment (yurishkuro) [reply extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java:225]: L225 replaces the builder, so they are only merged if jaeger-baggage header is first in the list, as it is in the unit test, moving this header to the end of the list will ignore uberctx entries, as far as I can tell.

Comment (carlosalberto) [reply extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java]: Verified that it works correctly by putting HEADER at the end (of the Map), i.e. L225 takes any **existing** builder and uses it if already exists.

File: context_prop/src/main/java/io/opentelemetry/context/propagation/TextMapPropagator.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (carlosalberto) [reply context_prop/src/main/java/io/opentelemetry/context/propagation/TextMapPropagator.java]: Done.

**Comment by carlosalberto** [comment]
@open-telemetry/java-approvers Please review the latest updates.

@yurishkuro Updated the code based on your feedback. In a follow up we can add the encoded-values case, if that works for you.

**Comment by carlosalberto** [comment]
@yurishkuro Also, there's the question on whether, upon **no existing baggage** at all in the carrier, we should put an empty `Baggage` in `Context` (for example, if `SpanContext` was extracted but no **baggage** was detected at alll).

**Review by yurishkuro — APPROVED** [review]

**Comment by yurishkuro** [comment]
> upon no existing baggage at all in the carrier, we should put an empty Baggage in Context

I would try to avoid that. Even if there is an immutable singleton `Baggage.empty()`, adding it into an immutable context still requires re-allocation of the context, which is a malloc for little benefit. If would just make `Baggage.fromContext(ctx)` return empty singleton when key is not present.

File: extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java (hunk 6 above)
Comment (anuraaga) [reply extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java]: Ah I meant an equal check on the length - if it starts with it, then the length is guarateed to be at least that length and the `<` here looks redundant

**Comment by jkwatson** [comment]
looks like this needs a rebase, @carlosalberto

**Review by anuraaga — APPROVED** [review]

**Review by jkwatson — APPROVED** [review]

File: extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java (hunk 6 above)
Comment (carlosalberto) [reply extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java]: Done.

**Comment by carlosalberto** [comment]
@jkwatson Rebased ;)
